tests/api2/test_attachment_querying.py

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+from pytest_dependency import depends
+
+sys.path.append(os.getcwd())
+
+from middlewared.test.integration.assets.nfs import nfs_share
+from middlewared.test.integration.assets.pool import dataset
+from middlewared.test.integration.utils import call, client
+
+
+PARENT_DATASET = 'test_parent'
+CHILD_DATASET = f'{PARENT_DATASET}/child_dataset'
+
+
+def test_attachment_with_child_path(request):
+    depends(request, ['pool_04'], scope='session')
+    with dataset(PARENT_DATASET) as parent_dataset:
+        parent_path = f'/mnt/{parent_dataset}'
+        assert call('pool.dataset.attachments_with_path', parent_path) == []
+
+        with nfs_share(parent_dataset):
+            attachments = call('pool.dataset.attachments_with_path', parent_path)
+            assert len(attachments) > 0, attachments
+            assert attachments[0]['type'] == 'NFS Share', attachments
+
+            with dataset(CHILD_DATASET) as child_dataset:
+                child_path = f'/mnt/{child_dataset}'
+                call('sharing.nfs.create', {'path': child_path})
+                attachments = call('pool.dataset.attachments_with_path', child_path)
+                assert len(attachments) > 0, attachments
```

Comment:
Will this test fail if you remove your patch? I am not nitpicking, I am genuinely trying to understand which type of incorrect behavior are you fixing.

Reply:
Yes it will - so the problem is that when we query attachments, we are tackling the following case:

If `/mnt/pool/smb` is a SMB share and if we were to see attachments of `/mnt/pool`, we would correctly retrieve that there is a SMB attachment involved.

Now if we check `/mnt/pool/smb` to see if there are attachments we would find that yes there is a SMB attachment involved.

However if we were to check any child of the configured path i.e `/mnt/pool/smb/child` middleware would not report any attachments for this path which is wrong as the path is being consumed by SMB.

This PR fixes/addresses that

Reply:
So with your changes if we try to delete `pool/smb/child` dataset, the system will destroy the share for `pool/smb`. Won't that be a little bit too much?

Reply:
I agree, that's very undesirable.